lib/puppet/configurer.rb

```diff
@@ -207,7 +207,7 @@ def run_internal(options)
       if catalog = prepare_and_retrieve_catalog_from_cache
         options[:catalog] = catalog
         @cached_catalog_status = 'explicitly_requested'
-        @environment = catalog.environment
+        @environment = catalog.environment unless Puppet[:strict_environment_mode]
```

Comment:
Having `strict_environment_mode` use the environment from the cached catalog seems reasonable. The run would always fail if `strict_environment_mode`, `environment`, and caching are all set and the environment and cached catalog don't match.

We should make that part of the documentation for it.